### DISCUSSION TITLE
How to change the time to answer (EA) - IP connection failed

### DISCUSSION THREAD

**Discussion by dima-sz** [eModbus/eModbus#281] [Q&A]
I use Modbus Client TCP.
According to my measurements, the response with an error (EA) occurs 3 seconds after the request, I can reduce this time or get an answer (E0) instead of (EA).

**Comment by Miq1** [comment]
EA means that the TCP connection could not be stablished. This is mostly caused by the server not accepting connections. The E0 TIMEOUT error is set only, if there was a connection established, but the server did not answer in time.

The time it takes to detect a broken or unrecoverable connection is set somewhere in the TCP core code. I am not aware of any user definable timeout value there.

So my advice would be to rather look for the reason the server is dropping the TCP connection and  being unresponsive afterwards.

**Comment by dima-sz** [comment]
I poll several devices in a cycle, the response time from connected devices is 15-30ms, let's say I poll 10 devices, which means the data from each device will be updated every 150-300ms. I turn off one device and on this device the polling cycle stops for 3 seconds, if two are turned off, then the delay is 6s, etc.

setTarget IP-1
addRequest

setTarget IP-2
addRequest

setTarget IP-3
addRequest

.....

how to avoid such big delay on disconnected devices

**Reply by Miq1** [reply]
Well, as I wrote: there is no additional waiting time in the eModbus lib. It simply is calling ``connect()`` from the core code, that seems to wait for 3s to detect a broken connection.

Another solution would be to parallelize requests, but that would require a lot of redesign in the lib.

**Comment by dima-sz** [comment]
I tried to create my own ModbusTCP client for each device and it works, but the number is limited Free Size heap, 5-6 maximum, maybe there are some other ways?

**Reply by Miq1** [reply]
With each client you will create a worker task with separate stack area, so you will run into resource shortages inevitably.

I am afraid there is no easy way at the moment to solve your problem.

**Reply by dima-sz** [reply]
It looks like it's not just my problem, maybe it affects everyone?

**Reply by Miq1** [reply]
Your polling interval seems quite short as compared to others I saw.
One idea I had: why not turning down the frequency of polls for a server as soon as the first EA error is encountered? If a connection could be established you would crank up the frequency again.

**Comment by bertmelis** [comment]
Doesn't `WiFiClient::setTimeout()` set a connect timeout? We can apply the "respond-timeout" to connecting.
Setting a very low timeout is not without risk though.

**Reply by dima-sz** [reply]
setTimeout() has no effect.

**Reply by Miq1** [reply]
It has, in fact, only but it cares about quiet times in an established connection only IMHO.

**Reply by bertmelis** [reply]
Is that true? Seems like a bug to me then.

WiFiClient::connect blocks on a call to 'select' (from lwip) with the defined timeout. Timeout can be set i nthe connect call of by using setTimeout.

**Reply by Miq1** [reply]
May be I am wrong. It is quite a while since I put that fact in my mental drawers.
